docs/config info.txt
######################
# Config Explanation #
######################

#############################################################################################################################################
BOT SECTION:

Bot nickname                                        =>  "nickname"                         : "minion"                (nick)
Bot name                                            =>  "name"                             : "Bello!"                (name)
Bot ident                                           =>  "ident"                            : "minion"                (ident)
Bot own modes                                       =>  "bot modes"                        : "+ix"                   (modes +* /-*)
#############################################################################################################################################
SERVER SECTION:

Servers to set: (server ip or hostname:port:plain/ssl:optional password)
eg. irc.dal.net:6667:plain
eg. irc.dal.net:6679:ssl
eg. irc.dal.net:6667:plain:server_password
(default port is 6667/plain/without password)

If the server does not connect to the first server,
it will go to the next server in the list and try to connect, etc.

	"servers": [
            "localhost:6667:plain",
            "localhost:6679:ssl:password",
            "other-server:7000"
        ],

How many times to try to connect to a given server  =>  "how many times connect to server" : 5                       (number)
Time interval between connections to a given server =>  "connect delay"                    : 6                       (seconds)
Show server message of the day                      =>  "show message of the day"          : true                    (true/false)
#############################################################################################################################################
OWNER SECTION:

Bot owner info                                      =>  "bot admin"                        : "minion <user@local>"   (info)
Owner hashed main password                          =>  "owner password"                   : "abc"                   (password)
When owner joins channel, show message on channel   =>  "owner message on join channel"    : true                    (true/false)
Show message when owner joins                       =>  "owner message"                    : "Bello my master!"      (message)
#############################################################################################################################################
PRIVILEGES SECTION:
   
Owner(s) Hosts                                      =>  "OWNER"                            : "nick!ident@hostname"   (nick!ident@host, ...)
Admin(s) Hosts                                      =>  "ADMIN"                            : "nick!ident@hostname"   (nick!ident@host, ...)
User(s) Hosts                                       =>  "USER"                             : "nick!ident@hostname"   (nick!ident@host, ...)
#############################################################################################################################################
USERSLEVELS SECTION:

                                                    =>  "OWNER"                            : 0                       (number)
                                                    =>  "ADMIN"                            : 1                       (number)
                                                    =>  "USER"                             : 999                     (number)
#############################################################################################################################################
RESPONSE SECTION:

Bot response option                                 =>  "bot response"                     : "notice"                (notice/channel/priv)
#############################################################################################################################################
AUTOMATIC SECTION:

Auto op users from auto op list                     =>  "auto op"                          : true                    (true/false)
Auto op list                                        =>  "auto op list"                     : ["nick!id@host", "..."] (nick!ident@host)
Automatic rejoin channel when kicked                =>  "auto rejoin"                      : true                    (true/false)
Keep channel modes                                  =>  "keep channel modes"               : true                    (true/false)
Try to keep/recover nick from config                =>  "keep nick"                        : true                    (true/false)
#############################################################################################################################################
CHANNEL SECTION:

Channel name to join                                =>  "channel"                          : "#minion"               (channel name)
Join channel when connected                         =>  "auto join"                        : true                    (true/false)
Channel modes                                       =>  "channel modes"                    : "nt"                    (modes)
Channel password if set                             =>  "channel key"                      : ""                      (key)
Channel topic to set                                =>  "channel topic"                    : "bello!"                (topic)
Protect topic                                       =>  "keep topic"                       : true,                   (true/false)
Give voice to everyone in channel on join new user  =>  "give voice users on join"         : false                   (true/false)
#############################################################################################################################################
COMMANDS SECTION:

Send raw commands when connected                    =>  "raw commands on start"            : ["", "", "..."]         (raw commands)
#############################################################################################################################################
MESSAGE SECTION:

Show in cli channel user messages                   =>  "show channel user messages"       : false                   (true/false)
Show in cli channel kicks messages                  =>  "show channel kicks messages"      : true                    (true/false)
Show in cli users private messages                  =>  "show private messages"            : false                   (true/false)
Show in cli users notice messages                   =>  "show users notice messages"       : true                    (true/false)
Show in cli users joins to channel                  =>  "show users join channel"          : true                    (true/false)
Show in cli users parts channel                     =>  "show users part channel"          : true                    (true/false)
Show in cli users quit messages                     =>  "show users quit messages"         : true                    (true/false)
Show in cli users invite messages                   =>  "show users invite messages"       : true                    (true/false)
Show in cli channel topic changes                   =>  "show topic changes"               : true                    (true/false)
Show in cli users nick changes                      =>  "show nick changes"                : true                    (true/false)
Show in cli plugin usage info                       =>  "show plugin usage info"           : true                    (true/false)
Show in cli ctcp messages                           =>  "show ctcp messages"               : true                    (true/false)
#############################################################################################################################################
IGNORE SECTION:

Users to ignore (list)                              =>  "users"                            : ["", "", "..."]         (nick!ident@host)
#############################################################################################################################################
BANS SECTION:

Ban list                                            =>  "ban list":                        : ["", "", "..."]         (*!*@*)       
#############################################################################################################################################
COMMAND SECTION:

Command prefix                                      =>  "command prefix"                   : "!"                     (prefix)
#############################################################################################################################################
CTCP SECTION:

Ctcp response                                       =>  "ctcp response"                    : true                    (true/false)
#############################################################################################################################################
DELAYS SECTION:
    
Bot channel messages delay time                     =>  "channel delay"                    : 1                       (true/false)
Bot private messages delay time                     =>  "private delay"                    : 1	                     (true/false)
Bot notice messages delay time                      =>  "notice delay"                     : 1	                     (true/false)
#############################################################################################################################################
LOGS SECTION:

Log                                                 =>  "logging"                          : true                    (true/false)
Log bot messages                                    =>  "log bot messages"                 : true                    (true/false)
Log server messages                                 =>  "log server messages"              : true                    (true/false)
Log ctcp messages                                   =>  "log ctcp messages"                : true                    (true/false)
Log notice messages                                 =>  "log notice messages"              : true                    (true/false)
Log channel messages                                =>  "log channel messages"             : true                    (true/false)
Log plugins usage messages                          =>  "log plugins usage messages"       : true                    (true/false)
Log raw messages"                                   =>  "log raw messages"                 : false                   (true/false)
#############################################################################################################################################
TIME SECTION:

Set timezone                                        =>  "timezone"                         : "Europe/Warsaw"         (timezone)
#############################################################################################################################################
FETCH SECTION:
    
Link to plugins                                     =>  "fetch server"                     : "https://..."           (link)
#############################################################################################################################################
PROGRAM SECTION:

Play sounds                                         =>  "play sounds"                      : true                    (true/false)
#############################################################################################################################################
DEBUG SECTION:

Show in cli raw messages                            =>  "show raw"                         : false                   (true/false)
Show in cli own messages in raw mode                =>  "show own messages in raw mode"    : false                   (true/false)
Show in cli debug messages                          =>  "show debug"                       : false                   (true/false)

DOCS/CHANGELOG.TXT

   v1.2.2 changes (17.12.2024):  
   - added flood protection (channel, privmsg, notice, ctcp messages)
   - new options in config ('FLOOD' section):
      "flood delay": 0.5         <- interval in float
      "channel flood": "bankick" <- action after channel flood, options: 'bankick', 'kick', 'warn'
      "privmsg flood": "ignore"  <- action after privmsg flood, options: 'ignore', 'warn'
      "notice flood":  "ignore"  <- action after notice flood, options: 'ignore', 'warn'
      "ctcp flood":    "ignore"  <- action after ctcp flood, options: 'ignore', 'warn'
      
      bankick = ban + kick in channel
      kick    = kick from channel
      ignore  = add user to ignore list (*!ident@hostname)
      warn    = send warning to user
   - fixed ignore list and corrected channel ignore
   - fixed topic loop in timer
// ------------------------------------------------------------------------------------------------------------------------------------------ 
   v1.2.1 changes (15.12.2024):
   - bug with increased CPU usage fixed
   - from now on, with each server it is possible to separately set the connection type (plain/ssl), port and password
      eg. "localhost:6667:plain", "localhost:6679:ssl:123", etc,.  
   - added checking configuration file for errors
   - added compression of logs from the previous day
      creates an '<date>.zip' file and deletes the folder
   - ctrl+c, ctrl+break handler corrected for windows only
   - bot can now ignore users, new structure in config:
      "IGNORE": {
        "users": [
           "nick!ident@hostname",
           "..."
      bot ignores all plugins, register, ctcp, privmsg, channel messages, invites, notice - commands/messages
   - new plugin 'ignore' Adds host to ignore list (!ignore <nick!ident@hostname>)
   - new ban list format:
      "ban list": [
          "nick!ident@hostname",
          "*!ident@hostname",
          "..."
   - new auto op list format:
      "auto op list": [
          "nick!ident@hostname",
          "..."
   - new cli argument '-x' create default configuration file (config.json)
   - updated 'config info.txt' file
   - updated php from 7.4.5 to 7.4.33
// ------------------------------------------------------------------------------------------------------------------------------------------
   v1.2.0 changes (10.12.2024):
   - ssl servers support
   - more servers can now be set
      "servers" :[
	           "irc.dal.net:6667",
		   "example.net:6669",
                   "..." (...)
                 ],
      if the server does not connect to the first server,
      it will go to the next server in the list and try to connect, etc.
   - new logs structure
   - logs now located in directories marked with date
   - separate logs for different events
      (ctcp.txt, notice.txt, server.txt, raw.txt, plugins.txt, bot.txt, #<channel>.<server>.txt)
   - new options in config:
      "ssl": true/false
      "log bot messages": true/false
      "log server messages": true/false
      "log ctcp messages": true/false
      "log notice messages": true/false
      "log channel messages": true/false
      "log plugins usage messages": true/false
      "log raw messages": true/false
      "show ctcp messages": true/false
   - bot now shows info about sending raw commands
   - "raw commands" in config changed
      from: "JOIN #test1;JOIN #test2"
      to  : "JOIN test1",
            "JOIN test2",
	    "..." etc,.
// ------------------------------------------------------------------------------------------------------------------------------------------
   v1.1.9 changes (01.12.2024):
   - added in config "owner message on join channel", if set to true and "owner message" is provided.
     This message will be shown if owner joins channel
   - added in config "give voice users on join channel", if bot got @ will give +v to everyone who joined channel
   - added in config "bot modes", you can set bot own mode eg. -i, +ix etc...
   - added in config "raw commands on start", you can set own raw commands eg. "JOIN #test1;JOIN #test2"
     (separated by ';')
   - added ctrl+c, ctrl+break handler
// ------------------------------------------------------------------------------------------------------------------------------------------
   v1.1.8 changes (29.11.2024):
   - config changed from ini to json format (config.ini => config.json)
   - deleted 'update check.bat' since we now checking it after start time
   - added config section MESSAGE and new options:
                          show channel user messages
                          show private messages
                          show users join channel
                          show users part channel
                          show users quit messages
                          show topic changes
                          show nick changes
                          show channel kicks messages
			  show plugin usage info
			  show users notice messages
			  show users invite messages
   - added "keep topic" (CHANNEL) in config
     bot can now protect topic if changed
   - added "channel topic" (CHANNEL) in config
     bot can now set topic if got @
   - removed plugins: pause, unpause, newnick, raw, server
   - log filename format changed <#channelname.servername.txt>
   - added time in raw messages
   - added reason in kick messages
   - fixed rejoining channel if key is set
   - fixed notice messages from user(s)
   - fixed ctcp messages
   - new sounds
   - some bot messages have been changed
   - code cleanup
// ------------------------------------------------------------------------------------------------------------------------------------------
   v1.1.7 changes (17.01.2023):
   - from now on, you can create 1-998 permissions + 'normal user' (999) and 'owner' (0)
   - The entitlement structure changed, As of now The owner directory can have any name

src/config.php
<?php
/* Copyright (c) 2013-2024, minions
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

//---------------------------------------------------------------------------------------------------------
 !in_array(PHP_SAPI, array('cli', 'cli-server', 'phpdbg')) ?
  exit('This script can\'t be run from a web browser. Use CLI terminal to run it<br>'.
       'Visit <a href="https://github.com/S3x0r/MINION/">this page</a> for more information.') : false;
//---------------------------------------------------------------------------------------------------------

function createDefaultConfigFile()
{
    $config =
<<<END
{
    "BOT": {
        "nickname": "minion",
        "name": "http://github.com/S3x0r/MINION",
        "ident": "minion",
        "bot modes": ""
    },
    "SERVER": {
        "servers": [
            "localhost:6667:plain",
            "localhost:6679:ssl:password",
            "other-server:7000"
        ],
        "how many times connect to server": 5,
        "connect delay": 6,
        "show message of the day": false
    },
    "OWNER": {
        "bot admin": "minion <user@localhost>",
        "owner password": "47a8f9b32ec41bd93d79bf6c1c924aaecaa26d9afe88c39fc3a638f420f251ed",
        "owner message on join channel": true,
        "owner message": "Bello my master!"
    },
    "PRIVILEGES": {
        "OWNER": "",
        "ADMIN": "",
        "USER": ""
    },
    "USERSLEVELS": {
        "OWNER": 0,
        "ADMIN": 1,
        "USER": 999
    },
    "RESPONSE": {
        "bot response": "notice"
    },
    "AUTOMATIC": {
        "auto op": true,
        "auto op list": [
            "user!ident@hostname"
        ],
        "auto rejoin": true,
        "keep channel modes": true,
        "keep nick": true
    },
    "CHANNEL": {
        "channel": "#minion",
        "auto join": true,
        "channel modes": "nt",
        "channel key": "",
        "channel topic": "bello!",
        "keep topic": true,
        "give voice users on join": false
    },
    "COMMANDS": {
        "raw commands on start": [
        ]
    },
    "FLOOD": {
        "flood delay": 0.5,
        "channel flood": "bankick",
        "privmsg flood": "ignore",
        "notice flood": "ignore",
        "ctcp flood": "ignore"
    },
    "MESSAGE": {
        "show channel user messages": false,
        "show channel kicks messages": true,
        "show private messages": false,
        "show users notice messages": true,
        "show users join channel": true,
        "show users part channel": true,
        "show users quit messages": true,
        "show users invite messages": true,
        "show topic changes": true,
        "show nick changes": true,
        "show plugin usage info": true,
        "show ctcp messages": true
    },
    "IGNORE": {
       "users": [
       ]
    },
    "BANS": {
        "ban list": [
            "nick!ident@hostname",
            "*!ident@hostname",
            "*!*@onlyhost"
        ]
    },
    "COMMAND": {
        "command prefix": "!"
    },
    "CTCP": {
        "ctcp response": true
    },
    "DELAYS": {
        "channel delay": 1,
        "private delay": 1,
        "notice delay": 1
    },
    "LOGS": {
        "logging": true,
        "log bot messages": true,
        "log server messages": true,
        "log ctcp messages": true,
        "log notice messages": true,
        "log channel messages": true,
        "log plugins usage messages": true,
        "log raw messages": false
    },
    "TIME": {
        "timezone": "Europe/Warsaw"
    },
    "FETCH": {
        "fetch server": "https://raw.githubusercontent.com/S3x0r/minion_repository_plugins/master"
    },
    "PROGRAM": {
        "play sounds": true
    },
    "DEBUG": {
        "show raw": false,
        "show own messages in raw mode": false,
        "show debug": false
    }
}
END;

    /* Save default config to file */
    saveToFile(getConfigFileName(), $config, 'w');

    if (!is_file(getConfigFileName())) {
        cliError('Cannot create default configuration file! Read-Only filesystem? Exiting!');
        winSleep(6);
    }
}
//---------------------------------------------------------------------------------------------------------
function isConfigProvidedFromArgument()
{   
    if (isset($_SERVER['argv'][1]) && $_SERVER['argv'][1] == '-c' && !empty($_SERVER['argv'][2]) && is_file($_SERVER['argv'][2])) {
        return true;
    } else {
             return false;
    }
}
//---------------------------------------------------------------------------------------------------------
function isServerProvidedFromArgument()
{
    if (isset($_SERVER['argv'][1]) && $_SERVER['argv'][1] == '-o' &&
       !empty($_SERVER['argv'][2]) && !empty($_SERVER['argv'][3]) && is_numeric($_SERVER['argv'][3])) {
        return true;
    } else {
             return false;
    }
}
//---------------------------------------------------------------------------------------------------------
function getConfigFileName()
{
    if (isConfigProvidedFromArgument()) {
        return $_SERVER['argv'][2];
    } else {
             return CONFIGFILE;
    }
}
//---------------------------------------------------------------------------------------------------------
function loadValueFromConfigFile($section, $value)
{
    $config = json_decode(file_get_contents(getConfigFileName()), true);

    if (isset($config[$section][$value])) { // dodano isset - sprawdzic!
        return $config[$section][$value];
    }
}
//---------------------------------------------------------------------------------------------------------
function saveValueToConfigFile($section, $option, $value)
{
    $config = json_decode(file_get_contents(getConfigFileName()), true);

    $replace = [$section => [$option => $value]];
    
    $newArray = array_replace_recursive($config, $replace);
    
    $newJsonData = file_put_contents(getConfigFileName(), json_encode($newArray, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));
}
//---------------------------------------------------------------------------------------------------------
function saveValueToListConfigFile($section, $option, $value)
{
    $config = json_decode(file_get_contents(getConfigFileName()), true);

    array_push($config[$section][$option], $value);

    $newJsonData = file_put_contents(getConfigFileName(), json_encode($config, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));
}
//---------------------------------------------------------------------------------------------------------
function checkIfConfigExists()
{
    if (is_file(getConfigFileName())) {
        return true;
    } else {
             return false;
    }
}
//---------------------------------------------------------------------------------------------------------
function checkIfConfigIsValid($configFile)
{
    $config = json_decode(file_get_contents($configFile), true);

    /* if valid json file, lets now check variables, etc */
    if (json_last_error() === JSON_ERROR_NONE) {
        $infoTxt    = 'Incorrect value in config: ';
        $infoTxtEnd = ' (expected true/false) Please correct it.';
  
        /* SERVER: how many times connect to server */
        if (empty($config['SERVER']['how many times connect to server']) xor !is_int($config['SERVER']['how many times connect to server'])) {
            cliConfigErr($infoTxt.'"how many times connect to server" (expected number value) Please correct it.');
        }
  
        /* SERVER: connect delay */
        if (empty($config['SERVER']['connect delay']) xor !is_int($config['SERVER']['connect delay'])) {
            cliConfigErr($infoTxt.'"connect delay" (expected number value) Please correct it.');
        }
  
        /* DELAYS: channel delay */
        if (empty($config['DELAYS']['channel delay']) xor !is_int($config['DELAYS']['channel delay'])) {
            cliConfigErr($infoTxt.'"channel delay" (expected number) Please correct it.');
        }
  
        /* DELAYS: private delay */
        if (empty($config['DELAYS']['private delay']) xor !is_int($config['DELAYS']['private delay'])) {
            cliConfigErr($infoTxt.'"private delay" (expected number) Please correct it.');
        }
  
        /* DELAYS: notice delay */
        if (empty($config['DELAYS']['notice delay']) xor !is_int($config['DELAYS']['notice delay'])) {
            cliConfigErr($infoTxt.'"notice delay" (expected number) Please correct it.');
        }

        /* FLOOD: flood delay */
        if (empty($config['FLOOD']['flood delay']) xor !is_float($config['FLOOD']['flood delay'])) {
            cliConfigErr($infoTxt.'"flood delay" (expected float) Please correct it.');
        }

        /* RESPONSE: bot response */
        if (empty($config['RESPONSE']['bot response']) xor !in_array($config['RESPONSE']['bot response'], array("channel", "notice", "priv"), true)) {
            cliConfigErr($infoTxt.'"bot response" (expected channel/notice/priv) Please correct it.');
        }

        /* FLOOD: channel flood */
        if (empty($config['FLOOD']['channel flood']) xor !in_array($config['FLOOD']['channel flood'], array("bankick", "kick", "warn"), true)) {
            cliConfigErr($infoTxt.'"channel flood" (expected bankick/kick/warn) Please correct it.');
        }

        /* FLOOD: privmsg flood */
        if (empty($config['FLOOD']['privmsg flood']) xor !in_array($config['FLOOD']['privmsg flood'], array("ignore", "warn"), true)) {
            cliConfigErr($infoTxt.'"privmsg flood" (expected ignore/warn) Please correct it.');
        }

        /* FLOOD: notice flood */
        if (empty($config['FLOOD']['notice flood']) xor !in_array($config['FLOOD']['notice flood'], array("ignore", "warn"), true)) {
            cliConfigErr($infoTxt.'"notice flood" (expected ignore/warn) Please correct it.');
        }

        /* FLOOD: ctcp flood */
        if (empty($config['FLOOD']['ctcp flood']) xor !in_array($config['FLOOD']['ctcp flood'], array("ignore", "warn"), true)) {
            cliConfigErr($infoTxt.'"ctcp flood" (expected ignore/warn) Please correct it.');
        }

        (empty($config['BOT']['nickname']))                                     ? cliConfigErr($infoTxt.'"nickname" Please fill in missing data!')       : false;
        (empty($config['BOT']['name']))                                         ? cliConfigErr($infoTxt.'"name" Please fill in missing data!')           : false;
        (empty($config['BOT']['ident']))                                        ? cliConfigErr($infoTxt.'"ident" Please fill in missing data!')          : false;
        (empty($config['SERVER']['servers'][0]))                                ? cliConfigErr($infoTxt.'"servers" Please fill in missing data!')        : false;
        (empty($config['COMMAND']['command prefix']))                           ? cliConfigErr($infoTxt.'"command prefix" Please fill in missing data!') : false;
        (empty($config['OWNER']['owner password']))                             ? cliConfigErr($infoTxt.'"owner password" Please fill in missing data!') : false;

        (emptyOrNoBool($config, 'SERVER'    , 'show message of the day'))       ? cliConfigErr($infoTxt.'"show message of the day"'.$infoTxtEnd)         : false;
        (emptyOrNoBool($config, 'OWNER'     , 'owner message on join channel')) ? cliConfigErr($infoTxt.'"owner message on join channel"'.$infoTxtEnd)   : false;
        (emptyOrNoBool($config, 'AUTOMATIC' , 'auto op'))                       ? cliConfigErr($infoTxt.'"auto op"'.$infoTxtEnd)                         : false;
        (emptyOrNoBool($config, 'AUTOMATIC' , 'auto rejoin'))                   ? cliConfigErr($infoTxt.'"auto rejoin"'.$infoTxtEnd)                     : false;
        (emptyOrNoBool($config, 'AUTOMATIC' , 'keep channel modes'))            ? cliConfigErr($infoTxt.'"keep channel modes"'.$infoTxtEnd)              : false;
        (emptyOrNoBool($config, 'AUTOMATIC' , 'keep nick'))                     ? cliConfigErr($infoTxt.'"keep nick"'.$infoTxtEnd)                       : false;
        (emptyOrNoBool($config, 'CHANNEL'   , 'auto join'))                     ? cliConfigErr($infoTxt.'"auto join"'.$infoTxtEnd)                       : false;
        (emptyOrNoBool($config, 'CHANNEL'   , 'keep topic'))                    ? cliConfigErr($infoTxt.'"keep topic"'.$infoTxtEnd)                      : false;
        (emptyOrNoBool($config, 'CHANNEL'   , 'give voice users on join'))      ? cliConfigErr($infoTxt.'"give voice users on join"'.$infoTxtEnd)        : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show channel user messages'))    ? cliConfigErr($infoTxt.'"show channel user messages"'.$infoTxtEnd)      : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show channel kicks messages'))   ? cliConfigErr($infoTxt.'"show channel kicks messages"'.$infoTxtEnd)     : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show private messages'))         ? cliConfigErr($infoTxt.'"show private messages"'.$infoTxtEnd)           : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show users notice messages'))    ? cliConfigErr($infoTxt.'"show users notice messages"'.$infoTxtEnd)      : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show users join channel'))       ? cliConfigErr($infoTxt.'"show users join channel"'.$infoTxtEnd)         : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show users part channel'))       ? cliConfigErr($infoTxt.'"show users part channel"'.$infoTxtEnd)         : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show users quit messages'))      ? cliConfigErr($infoTxt.'"show users quit messages"'.$infoTxtEnd)        : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show users invite messages'))    ? cliConfigErr($infoTxt.'"show users invite messages"'.$infoTxtEnd)      : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show topic changes'))            ? cliConfigErr($infoTxt.'"show topic changes"'.$infoTxtEnd)              : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show nick changes'))             ? cliConfigErr($infoTxt.'"show nick changes"'.$infoTxtEnd)               : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show plugin usage info'))        ? cliConfigErr($infoTxt.'"show plugin usage info"'.$infoTxtEnd)          : false;
        (emptyOrNoBool($config, 'MESSAGE'   , 'show ctcp messages'))            ? cliConfigErr($infoTxt.'"show ctcp messages"'.$infoTxtEnd)              : false;
        (emptyOrNoBool($config, 'CTCP'      , 'ctcp response'))                 ? cliConfigErr($infoTxt.'"ctcp response"'.$infoTxtEnd)                   : false;
        (emptyOrNoBool($config, 'LOGS'      , 'logging'))                       ? cliConfigErr($infoTxt.'"logging"'.$infoTxtEnd)                         : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log bot messages'))              ? cliConfigErr($infoTxt.'"log bot messages"'.$infoTxtEnd)                : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log server messages'))           ? cliConfigErr($infoTxt.'"log server messages"'.$infoTxtEnd)             : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log ctcp messages'))             ? cliConfigErr($infoTxt.'"log ctcp messages"'.$infoTxtEnd)               : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log notice messages'))           ? cliConfigErr($infoTxt.'"log notice messages"'.$infoTxtEnd)             : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log channel messages'))          ? cliConfigErr($infoTxt.'"log channel messages"'.$infoTxtEnd)            : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log plugins usage messages'))    ? cliConfigErr($infoTxt.'"log plugins usage messages"'.$infoTxtEnd)      : false;
        (emptyOrNoBool($config, 'LOGS'      , 'log raw messages'))              ? cliConfigErr($infoTxt.'"log raw messages"'.$infoTxtEnd)                : false;
        (emptyOrNoBool($config, 'PROGRAM'   , 'play sounds'))                   ? cliConfigErr($infoTxt.'"play sounds"'.$infoTxtEnd)                     : false;
        (emptyOrNoBool($config, 'DEBUG'     , 'show raw'))                      ? cliConfigErr($infoTxt.'"show raw"'.$infoTxtEnd)                        : false;
        (emptyOrNoBool($config, 'DEBUG'     , 'show own messages in raw mode')) ? cliConfigErr($infoTxt.'"show own messages in raw mode"'.$infoTxtEnd)   : false;
        (emptyOrNoBool($config, 'DEBUG'     , 'show debug'))                    ? cliConfigErr($infoTxt.'"show debug"'.$infoTxtEnd)                      : false;
     } else {
              /* if invalid json format */
              cliNoLog('');
              cliBot('Config file ('.getConfigFileName().') contains JSON syntax errors, Correct errors in config!');
              winSleep(7);
              exit;
     }
}
//---------------------------------------------------------------------------------------------------------
function cliConfigErr($info)
{
    cliError($info);
    winSleep(7);
}
//---------------------------------------------------------------------------------------------------------
function emptyOrNoBool($config, $section, $value)
{
    $data = strtolower($config[$section][$value]);

    if (empty($data) xor !in_array($data, array("true", "false", "1", "0"), true)) {
        return true;
    }
}
